graph.go

```diff
@@ -4,193 +4,166 @@
-	Graph
-	// Successors gives the nodes connected by OUTBOUND edges.
-	// If the graph is an undirected graph, this set is equal to Predecessors.
-	Successors(Node) []Node
-
-	// EdgeTo returns an edge between node and successor such that
-	// From returns node and To returns successor, if no
-	// such edge exists, this function returns nil.
-	EdgeTo(node, successor Node) Edge
-
-	// Predecessors gives the nodes connected by INBOUND edges.
-	// If the graph is an undirected graph, this set is equal to Successors.
-	Predecessors(Node) []Node
-}
+	// Nodes returns all the nodes in the graph.
+	Nodes() []Node
 
-// Returns all undirected edges in the graph
-type EdgeLister interface {
-	EdgeList() []Edge
-}
+	// From returns all nodes that can be reached from
+	// the given node.
+	From(Node) []Node
 
-type EdgeListGraph interface {
-	Graph
-	EdgeLister
-}
+	// HasEdge returns whether an edge exists between
+	// nodes x and y without considering direction.
+	HasEdge(x, y Node) bool
 
-// Returns all directed edges in the graph.
-type DirectedEdgeLister interface {
-	DirectedEdgeList() []Edge
+	// Edge returns the edge between nodes u and v when
+	// the nodes returned by From(u) include v.
+	Edge(u, v Node) Edge
 }
 
-type DirectedEdgeListGraph interface {
+// Undirected is an undirected graph.
+type Undirected interface {
 	Graph
-	DirectedEdgeLister
-}
 
-// A Graph that implements Coster has an actual cost between adjacent nodes, also known as a
-// weighted graph. If a graph implements coster and a function needs to read cost (e.g. A*),
-// this function will take precedence over the Uniform Cost function (all weights are 1) if "nil"
-// is passed in for the function argument.
-//
-// If the argument is nil, or the edge is invalid for some reason, this should return math.Inf(1)
-type Coster interface {
-	Cost(Edge) float64
+	// EdgeBetween returns the edge between nodes u and v.
+	EdgeBetween(u, v Node) Edge
 }
 
-type CostGraph interface {
-	Coster
+// Directed is a directed graph.
+type Directed interface {
 	Graph
+
+	// EdgeFromTo returns the edge leading from u to v.
+	EdgeFromTo(u, v Node) Edge
+
+	// To returns all nodes that can be lead to the
+	// given node.
+	To(Node) []Node
 }
 
-type CostDirectedGraph interface {
-	Coster
-	DirectedGraph
+// EdgeLister wraps the Edges method.
+type EdgeLister interface {
+	Edges() []Edge
 }
 
-// A graph that implements HeuristicCoster implements a heuristic between any two given nodes.
-// Like Coster, if a graph implements this and a function needs a heuristic cost (e.g. A*), this
-// function will take precedence over the Null Heuristic (always returns 0) if "nil" is passed in
-// for the function argument. If HeuristicCost is not intended to be used, it can be implemented as
-// the null heuristic (always returns 0).
-type HeuristicCoster interface {
-	// HeuristicCost returns a heuristic cost between any two nodes.
-	HeuristicCost(n1, n2 Node) float64
+// Weighter wraps the Weight method.
```

Comment:
Actually, this is not true. The edges may have a weight, just that the graph does not know about them.